File: goatx402-sdk-server-go/signature.go
package goatx402

import (
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"sort"
	"strings"
	"time"
)

// calculateSignature computes HMAC-SHA256 signature for API authentication
//
// Algorithm:
// 1. Sort all parameters by key in ASCII order
// 2. Concatenate as key1=value1&key2=value2 format
// 3. Compute HMAC-SHA256 using the API secret
// 4. Return hexadecimal string
func calculateSignature(params map[string]string, secret string) string {
	// Remove sign field if present
	delete(params, "sign")

	// Get sorted keys
	keys := make([]string, 0, len(params))
	for k := range params {
		if params[k] != "" { // Ignore empty values
			keys = append(keys, k)
		}
	}
	sort.Strings(keys)

	// Build signature string
	parts := make([]string, 0, len(keys))
	for _, k := range keys {
		parts = append(parts, fmt.Sprintf("%s=%s", k, params[k]))
	}
	signStr := strings.Join(parts, "&")

	// Compute HMAC-SHA256
	h := hmac.New(sha256.New, []byte(secret))
	h.Write([]byte(signStr))
	return hex.EncodeToString(h.Sum(nil))
}

func generateRequestNonce() string {
	var buf [16]byte
	if _, err := rand.Read(buf[:]); err == nil {
		return hex.EncodeToString(buf[:])
	}
	return fmt.Sprintf("%d", time.Now().UnixNano())
}

File: goatx402-sdk-server-ts/src/signature.ts
/**
 * HMAC-SHA256 signature utilities for GoatX402 API authentication
 */

import { createHmac, randomBytes, randomUUID } from 'crypto'

/**
 * Calculate HMAC-SHA256 signature for API request
 *
 * Algorithm:
 * 1. Sort all parameters by key in ASCII order
 * 2. Concatenate as key1=value1&key2=value2 format
 * 3. Encrypt using HMAC-SHA256 algorithm
 * 4. Return hexadecimal string
 */
export function calculateSignature(
  params: Record<string, string>,
  secret: string
): string {
  // Exclude sign field if present
  const filteredParams = { ...params }
  delete filteredParams.sign

  // Sort keys alphabetically
  const sortedKeys = Object.keys(filteredParams)
    .filter((k) => filteredParams[k] !== '')
    .sort()

  // Build signature string
  const signStr = sortedKeys.map((k) => `${k}=${filteredParams[k]}`).join('&')

  // Compute HMAC-SHA256
  const signature = createHmac('sha256', secret).update(signStr).digest('hex')

  return signature
}

/**
 * Generate authentication headers for API request
 */
export function signRequest(
  params: Record<string, unknown>,
  apiKey: string,
  apiSecret: string
): {
  'X-API-Key': string
  'X-Timestamp': string
  'X-Nonce': string
  'X-Sign': string
} {
  // Convert all params to strings
  const strParams: Record<string, string> = {}
  for (const [k, v] of Object.entries(params)) {
    if (v !== undefined && v !== null) {
      strParams[k] = String(v)
    }
  }

  // Add authentication parameters
  const timestamp = Math.floor(Date.now() / 1000).toString()
  const nonce =
    typeof randomUUID === 'function'
      ? randomUUID()
      : `${Date.now().toString(36)}-${randomBytes(12).toString('hex')}`
  strParams.api_key = apiKey
  strParams.timestamp = timestamp
  strParams.nonce = nonce

  // Calculate signature
  const sign = calculateSignature(strParams, apiSecret)

  return {
    'X-API-Key': apiKey,
    'X-Timestamp': timestamp,
    'X-Nonce': nonce,
    'X-Sign': sign,
  }
}
